source/simulator/stokes_matrix_free.cc

```diff
@@ -46,15 +46,142 @@ namespace aspect
     /**
      * Here we define the function(s) to make no normal flux boundary constraints for
      * MG levels.
-     * TODO: Implement for spherical meshes.
      */
     namespace TangentialBoundaryFunctions
     {
+      template <int dim, int spacedim>
+      void compute_no_normal_flux_constraints_shell(const DoFHandler<dim,spacedim> &dof_handler,
+                                                    const MGConstrainedDoFs        &mg_constrained_dofs,
+                                                    const Mapping<dim> &mapping,
+                                                    const unsigned int level,
+                                                    const unsigned int first_vector_component,
+                                                    const std::set<types::boundary_id> &boundary_ids,
+                                                    ConstraintMatrix &constraints)
+      {
+        // TODO: This is a simplification of compute_no_normal_flux_constraints() from deal.II.
+        // The differences are:
+        // - It works on a specific level so we can ignore hanging nodes
+        // - We use the normal vector given by the manifold (instead of averaging surface vectors)
+        //
+        // This should go into deal.II at some point, but it is too specific at this point.
+
+        IndexSet refinement_edge_indices = mg_constrained_dofs.get_refinement_edge_indices(level);
```

Comment:
For GMG, the DoFs along the refinement edge of a level (i.e., DoFs that lie between cells of different refinement levels, and therefore exist on the current MG level and the level below) are set to zero before smoothing.  This is due to the fact that, since this DoF also exists on the level coarser, we only want to smooth there so as not to repeat work. Therefore we do not want to prescribe no-normal-flux constraints on these DoFs on this level. Instead, we let the MG transfer set the DoF to zero.

Reply:
added a comment (below where it is used)